### ISSUE TITLE
Не понятно как использовать модель в tritonserver

### ISSUE THREAD

**Issue by slavonnet** [voicekit-team/T-one#15]
Модель загружена но input у нее не понятные

```json
{"name":"streaming_acoustic","versions":["1"],"platform":"tensorrt_plan","inputs":[{"name":"signal","datatype":"INT32","shape":[-1,2400,1]},{"name":"state","datatype":"FP16","shape":[-1,219729]}],"outputs":[{"name":"logprobs","datatype":"FP32","shape":[-1,10,35]},{"name":"state_next","datatype":"FP16","shape":[-1,219729]}]}
```

хоть минимальный пример дали бы как в нее подавать данные. В описании Triton Server про аудио модели и как их использовать нет  ничего

**Comment by Ar4ikov** [comment]
Добрый день! 
Ниже делюсь своим опытом подключения акустической модели, который был реализован через ~~боль~~вайбкодинг с собственными поисками.
Для работы с tritonserver необходима любая его клиентская имплементация. 
Наверное, самый правильный путь работы - использовать tritonclient под gRPC.
https://pypi.org/project/tritonclient/

```bash
pip install tritonclient[grpc]
```

Примеры исполнений клиентов есть в репо на Гитхабе:
https://github.com/triton-inference-server/client/tree/main

gRPC client: https://github.com/triton-inference-server/client/blob/main/src/python/examples/grpc_client.py
Simple Stream Sequence GRPC Infer: https://github.com/triton-inference-server/client/blob/main/src/python/examples/simple_grpc_sequence_stream_infer_client.py

Аудиопоток в сыром виде (PCM) необходимо перевести в int32. Вариантов много, ниже - небольшой пример:
```python
if audio_bytes and len(audio_bytes) % 2 == 0:
  try:
      pcm = np.frombuffer(audio_bytes, dtype=np.int16)
      if pcm.size:
          audio_data = pcm.astype(np.float32) / 32768.0
  except ValueError:
      audio_data = None
  
if audio_data is None:
  try:
      audio_io = io.BytesIO(audio_bytes)
      audio_data, detected_rate = sf.read(audio_io)
      if detected_rate:
          source_rate = int(detected_rate)
      if audio_data.ndim > 1:
          audio_data = audio_data.mean(axis=1)
      audio_data = audio_data.astype(np.float32)
  except Exception:
      _LOGGER.debug("Fallback to raw PCM parsing for Triton backend")
      if not audio_bytes or len(audio_bytes) % 2 != 0:
          return np.zeros((0,), dtype=np.int32)
      pcm = np.frombuffer(audio_bytes, dtype=np.int16)
      if pcm.size == 0:
          return np.zeros((0,), dtype=np.int32)
      audio_data = pcm.astype(np.float32) / 32768.0
```
Важно еще аудио ресемплировать в нужный для T-One `SAMPLE_RATE`
`torchaudio` или `librosa` - любой вариант за вами. 
Чтобы использовать `SAMPLE_RATE` как в t_one модели, можно импортировать `StreamingCTCModel` из `t_one.onnx_wrapper`:
```python
from tone.onnx_wrapper import StreamingCTCModel
SAMPLE_RATE = StreamingCTCModel.SAMPLE_RATE
```

Как в целом и остальную часть конфигурации:
```python
STATE_SIZE = StreamingCTCModel.STATE_SIZE
CHUNK_SAMPLES = StreamingCTCModel.AUDIO_CHUNK_SAMPLES
FRAME_SIZE = StreamingCTCModel.FRAME_SIZE
TIME_BIAS = StreamingCTCModel.MEAN_TIME_BIAS
```

Далее при помощи документации tritonclient нужно инициировать `GrpcInferInput`, `GrpcInferOutput`.
Входных данных 2 - `signal` и `state`, выходных тоже 2 - `logprobs` и `state_next` - вероятностный CTC выход модели для букф под логарифмом и скрытое состояние акустической модели для передачи в контекст к обработке следующего аудиочанка в последовательности.

```python
signal_input = GrpcInferInput("signal", (1, CHUNK_SAMPLES, 1), "INT32")
signal_input.set_data_from_numpy(chunk.reshape(1, CHUNK_SAMPLES, 1).astype(np.int32))

state_input = GrpcInferInput("state", state.model_state.shape, "FP16")
state_input.set_data_from_numpy(state.model_state.astype(np.float16))
```

```python
_requested_outputs_grpc = [
    GrpcInferRequestedOutput("logprobs"),
    GrpcInferRequestedOutput("state_next"),
]
```

Инициализировать сам GRPC клиент:
```python
grpc_client = GrpcInferenceServerClient(
    url=self.server_url,
    verbose=False,
    ssl=self._use_ssl,
)
```

Опционально: добавить хелсчек на готовность модели по /v2/health/ready + готовность модели:
```python
grpc_client.is_server_live()
grpc_client.is_server_ready()
grpc_client.is_model_ready(model_name)

# стандартное имя модели: streaming_acoustic или streaming_acoustic_trt
```

А дальше сделать вызов обработки модели:
```python
try:
    response = self.grpc_client.infer(
        model_name=self.model_name,
        inputs=[signal_input, state_input],
        outputs=self._requested_outputs_grpc,
        timeout=self._request_timeout,
    )
except triton_utils.InferenceServerException as exc:  # type: ignore[attr-defined]
    _LOGGER.error("Triton gRPC inference failed", error=str(exc))
    raise

logprobs = response.as_numpy("logprobs")
state_next = response.as_numpy("state_next")
```

Требуются дополнительно некоторые преобразования (например, state_next в `np.float16`)
Еще важно заметить - можно собирать несколько потоков данных в батч и оптравлять их в `TritonServer`.
Для этого нужно подготовить корректно работу по потокам для `state_next` данных, передавая их в одном едином входе через BATCH_SIZE (шейп тензора в самом начале), аналогично поступить и с `signals`

Модель может быть скомпилирована для работы сразу с батчем данных, при сборке указывается batch_size. В целом, на размер модели это влияет не сильно. Модель для работы с 256 батчами аудио чанков за раз стандартно занимает ~922 MB VRAM. 

Главное - если есть сразу несколько чанков одного потока аудио, в батч запихнуть такое не выйдет. Т.к. важно при работе с потоком обрабатывать последовательно каждый чанк самого потока, попытка завернуть в батч, скажем, заготовленный в очереди поток из 256 последовательных чанков по 300 мс приведет к тому, что результат транскрибации модели будет некорректным (по сути нарушаем логический принцип - последовательные данные передаем за один прогон, как будто знаем заранее все скрытые состояния). 

Последний шаг - как в коде библиотеки `tone` подключить SplitterLogProbs и нужный вам Decoder. Beam Search будет потреблять больше ресурсов, чем `greedy`, в т.ч. будет работать медленнее, однако, результат с `beam_search` в моих примерах получался точнее. Вероятно, метрики WER в документации были достигнуты использованием `beam_search` декодера. Однако, в блокноте файнтюна метрики на датасете Вихря ToneBooks были получены от `greedy` декодера:
> Training will take a couple of hours depending on the GPU used. During training train and validation loss will be printed along with WER of greedy predictions of the model on the validation set.

Выбор всегда остается за вами: скорость или качество. Можно использовать `greedy` и зафайнтюнить модель под данные внутри компании под свой конкретный домен.

Надеюсь, что смог помочь вам разобраться!
А команду очень сильно прошу обновить документацию и примеры использования для tritonserver <3 @sxdxfan 
Моделька крутая, хорошо обучили конформер, только вот инфру вокруг нее построить тяжеловато. Но возможно, и тогда всё полетит! 🤗

**Comment by DenDiv** [comment]
@slavonnet, добавили [инструкцию](https://github.com/voicekit-team/T-one/blob/main/examples/triton_request_example.ipynb) по тому как слать запросы в тритон, если есть ещё вопросы, переоткрой issue. @Ar4ikov, спасибо за развернутый ответ
